common/src/main/java/com/microsoft/identity/common/internal/result/MsalBrokerResultAdapter.java

```diff
@@ -182,12 +183,112 @@ public BaseException getBaseExceptionFromBundle(@NonNull final Bundle resultBund
+        }
+
+        if(exceptionType.equalsIgnoreCase(UserCancelException.sName)){
+            Logger.warn(TAG, "Received a User cancelled exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = new UserCancelException();
+        }
+
+        if(exceptionType.equalsIgnoreCase(ClientException.sName)){
+            Logger.warn(TAG, "Received a ClientException exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = new ClientException(
+                    brokerResult.getErrorCode(),
+                    brokerResult.getErrorMessage()
+            );
+        }
+
+        if(exceptionType.equalsIgnoreCase(ArgumentException.sName)){
+            Logger.warn(TAG, "Received a Argument exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = new ArgumentException(
+                    ArgumentException.ACQUIRE_TOKEN_OPERATION_NAME,
+                    brokerResult.getErrorCode(),
+                    brokerResult.getErrorMessage()
+            );
+        }
+
+        // Default to ClientException if null
+        if(baseException == null){
+            Logger.warn(TAG, " Exception type is unknown : " + exceptionType
+                    + brokerResult.getErrorCode()
+            );
+            baseException = new ClientException(
+                    brokerResult.getErrorCode(),
+                    brokerResult.getErrorMessage()
+            );
+        }
+
+        baseException.setCliTelemErrorCode(brokerResult.getCliTelemErrorCode());
+        baseException.setCliTelemSubErrorCode(brokerResult.getCliTelemSubErrorCode());
+        baseException.setCorrelationId(brokerResult.getCorrelationId());
+        baseException.setSpeRing(brokerResult.getSpeRing());
+        baseException.setRefreshTokenAge(brokerResult.getRefreshTokenAge());
+
+        return baseException;
+    }
+
+
+    /**
+     * Method to get the right base exception based on error codes.
+     * Note : In newer versions of Broker, exception type will be sent and is used to determine the right exception.
+     *
+     * This method is to support legacy broker versions (3.1.8 or below)
+     * @return BaseException
+     */
+    private BaseException getBaseExceptionFromErrorCodes(@NonNull final BrokerResult brokerResult){
         final String errorCode = brokerResult.getErrorCode();
+        final BaseException baseException;
 
         if (AuthenticationConstants.OAuth2ErrorCode.INTERACTION_REQUIRED.equalsIgnoreCase(errorCode) ||
-                AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT.equalsIgnoreCase(errorCode)) {
+                AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT.equalsIgnoreCase(errorCode) ||
+                ErrorStrings.INVALID_BROKER_REFRESH_TOKEN.equalsIgnoreCase(errorCode)) {
```

Comment:
`ErrorStrings.NO_TOKENS_FOUND` is another case for UiRequiredException

Reply:
Thanks for pointing out.